src/feature_builder.hpp

```diff
@@ -15,88 +14,181 @@
+    }
+
+    void linestring_point(vtzero::point const& pt)
+    {
+        CoordinateType x = pt.x * zoom_factor_ - dx_;
+        CoordinateType y = pt.y * zoom_factor_ - dy_;
+        geom_.back().emplace_back(x, y);
+    }
+
+    void linestring_end() {}
+
+    geom_type& geom_;
+    int dx_;
+    int dy_;
+    int zoom_factor_;
+};
+
+template <typename CoordinateType>
+struct polygon_ring
+{
+
+    polygon_ring()
+        : ring(), type(vtzero::ring_type::invalid)
+    {
+    }
+
+    mapbox::geometry::linear_ring<CoordinateType> ring;
+    vtzero::ring_type type;
+};
+
+template <typename CoordinateType>
+struct polygon_handler
+{
+    using geom_type = std::vector<polygon_ring<CoordinateType>>;
+    polygon_handler(geom_type& geom, int dx, int dy, int zoom_factor)
+        : geom_(geom),
+          dx_(dx),
+          dy_(dy),
+          zoom_factor_(zoom_factor) {}
+
+    void ring_begin(std::uint32_t count)
+    {
+        geom_.emplace_back();
+        geom_.back().ring.reserve(count);
+    }
+
+    void ring_point(vtzero::point const& pt)
+    {
+        CoordinateType x = pt.x * zoom_factor_ - dx_;
+        CoordinateType y = pt.y * zoom_factor_ - dy_;
+        geom_.back().ring.emplace_back(x, y);
+    }
+
+    void ring_end(vtzero::ring_type type)
+    {
+        geom_.back().type = type;
+    }
+
+    geom_type& geom_;
+    int dx_;
+    int dy_;
+    int zoom_factor_;
+};
+
+} // namespace detail
+
+template <typename CoordinateType>
+struct overzoomed_feature_builder
 {
     using coordinate_type = CoordinateType;
-    feature_builder_visitor(vtzero::layer_builder& layer_builder,
-                            mapbox::geometry::box<coordinate_type> const& bbox,
-                            vtzero::feature const& feature)
+    overzoomed_feature_builder(vtzero::layer_builder& layer_builder,
+                               mapbox::geometry::box<coordinate_type> const& bbox,
+                               int dx, int dy, int zoom_factor)
         : layer_builder_{layer_builder},
           bbox_{bbox},
-          feature_{feature} {}
+          dx_{dx},
+          dy_{dy},
+          zoom_factor_{zoom_factor} {}
 
     template <typename FeatureBuilder>
-    void finalize(FeatureBuilder& builder)
+    void finalize(FeatureBuilder& builder, vtzero::feature const& feature)
     {
         // add properties
-        feature_.for_each_property([&builder](vtzero::property const& p) {
+        feature.for_each_property([&builder](vtzero::property const& p) {
             builder.add_property(p);
             return true;
         });
         builder.commit();
     }
 
-    void operator()(mapbox::geometry::point<coordinate_type> const& pt)
+    void apply_geometry_point(vtzero::feature const& feature)
     {
+        mapbox::geometry::multi_point<CoordinateType> multi_point;
+        vtzero::decode_point_geometry(feature.geometry(), detail::point_handler<coordinate_type>(multi_point, dx_, dy_, zoom_factor_));
```

Comment:
What do you think about passing the `bbox` down into the `detail::point_handler`? If it ends up filters points then of course the `reserve` would be slightly in excess but the benefit would be that you could avoid the allocation required to build up the `std::vector<mapbox::geometry::point<coordinate_type>> result;` below. Instead the `multi_point` would have all the points you'd need to pass along and then you could:

```c++
vtzero::add_points_from_container(multi_point);
```

Reply:
@springmeyer I tried passing bbox ^, only creating single points and even completely commenting out point handling code - still node-mapnik is faster. hmm..

Reply:
ok, issue resolved and fixed - it was :apple: vs :tangerine: benching. Now, benching results look good again. @millzpaugh - would be great if you can go throw benching rules with fresh eyes and see if anything else we might need to fix.

Reply:
@artemp I'm curious why we switched the benching script to have `buffer_size=0` - I originally added this to `buffer_size=128` because that's the node-mapnik default.